Rewrite for SL4

FichteFoll opened the pull request:
Conform to SL4 changes, using better integration and a couple bugfixes. Also uses rustc's JSON error format.

All old settings have been removed and replaced by new ones.

I mean, it is pretty much a complete rewrite.

If you're not looking forward to maintaining the package anymore, I can do so. I maintain two other linter plugins already and followed the SL4 rewrite, allowing me to know where to make changes.
If you agree, I would move the fork over to my main account and replace the entry in Package Control's default channel. Or you can transfer the repo over, if you like.

Fixes #33, fixes #25, fixes #24 (by using SL4's `working_dir` setting), fixes #21 (through the custom command setting), fixes #5
Related: #26.

Probably needs some further testing as I'm sure I haven't run into all the corner cases with my very simple test setup.

FichteFoll committed:
Complete rewrite

Conform to SL4 changes, better integration and bugfixes.

I mean, it is pretty much a complete rewrite.

mandeep commented:
@FichteFoll will this plugin still work for those with SL3? I found that SL4 broke a lot of plugins, so I'm still on SL3 myself.

FichteFoll commented:
No, it will not work with SL3 and I have no intention to make it work.

FichteFoll commented:
Actually, I think I'll be using Rust Enhanced instead. It's better suited for the toolchain and the inline phantoms are quite helpful since they include all information from rustc rather than just the "message" summary line.

oschwald commented:
Thanks! I'd be happy to merge and release this if I could get some help testing it. I no longer actively use Sublime.

FichteFoll commented:
Well, the steps are pretty much as they were before.

Install SublimeLinter, clone my branch into the Packages dir and open a rust file or cargo project in ST (that means the folder). It should Just Work, assuming `cargo` and/or `rustc` are on your PATH.

mandeep commented:
I'm not really sure I agree with merging this. A lot of SL3 plugins haven't seen updates so upgrading this plugin to SL4 will break everyone on SL3.

FichteFoll commented:
Can you name a couple old plugins that don't work with sl4? We should rather update those to be compatible rather than hold back updates from other plugins because of this.

Also, does this plugin work for you at all in its current state?

mandeep commented:
I remember sublimelinter-gcc being a problem, but I haven't tried it in awhile so the problems may be fixed. 

This plugin currently works fine for me. Are you seeing issues?

FichteFoll commented:
It didn't work for individual files using rustc anymore because of a change in the error format over a year ago. That's when I added json parsing for myself via https://github.com/oschwald/SublimeLinter-contrib-rustc/issues/26#issuecomment-259843971.

However, that also didn't work anymore when I tried it earlier this month, so I went ahead and rewrote the entire thing to support the new SL4 features such as highlighting the entire region reported by rustc (even over multiple lines) and generally being more stable due to using the JSON output format.

Anyway, due to how well RustEnhanced integrates the output messages of rustc (considering they usually consist of multiple lines and notes), I'm not using this package anymore.

Edit: Just as an update from the future (2023), I use LSP and LSP-rust-analyzer for Rust nowadays, along with the built-in Rust syntax.

eread commented:
> It didn't work for individual files using rustc anymore because of a change in the error format over a year ago. That's when I added json parsing for myself via [#26 (comment)](https://github.com/oschwald/SublimeLinter-contrib-rustc/issues/26#issuecomment-259843971).
> 
> However, that also didn't work anymore when I tried it earlier this month, so I went ahead and rewrote the entire thing to support the new SL4 features such as highlighting the entire region reported by rustc (even over multiple lines) and generally being more stable due to using the JSON output format.
> 
> Anyway, due to how well RustEnhanced integrates the output messages of rustc (considering they usually consist of multiple lines and notes), I'm not using this package anymore.
> 
> Edit: Just as an update from the future (2023), I use LSP and LSP-rust-analyzer for Rust nowadays, along with the built-in Rust syntax.

@FichteFoll Is this PR still worth merging for folks that do use SublimeLinter for things?

FichteFoll commented:
Probably? As I mentioned in the cited comment, I'm not using this anymore, so I suggest to try it out yourself if it still works.

Can't be worse than what this repo currently does, though.
